sdk/storage/Azure.Storage.DataMovement/src/JobPartInternal.cs

```diff
@@ -178,26 +182,75 @@ internal abstract class JobPartInternal
             }
 
             Length = length;
+            _chunkTasks = new List<Task<bool>>();
+            _chunkTaskSources = new List<TaskCompletionSource<bool>>();
         }
 
         public void SetQueueChunkDelegate(QueueChunkDelegate chunkDelegate)
         {
             QueueChunk = chunkDelegate;
         }
 
+        /// <summary>
+        /// Queues the task to the main chunk channel and also appends the tracking
+        /// completion source to the task. So we know the state of each chunk especially
+        /// when we're looking to stop/pause the job part.
+        /// </summary>
+        /// <returns></returns>
+        public async Task QueueChunkToChannelAsync(Task chunkTask)
+        {
+            // Attach TaskCompletionSource
+            TaskCompletionSource<bool> chunkCompleted = new TaskCompletionSource<bool>(
+                false,
+                TaskCreationOptions.RunContinuationsAsynchronously);
+            _chunkTaskSources.Add(chunkCompleted);
+            _chunkTasks.Add(chunkCompleted.Task);
+
+            await QueueChunk(
+                async () =>
+                {
+                    await chunkTask.ConfigureAwait(false);
+                    chunkCompleted.SetResult(true);
+                    await CheckAndUpdateCancellationStatusAsync().ConfigureAwait(false);
```

Comment:
@jalauzon-msft CheckAndUpdateCancellationStatusAsync is  also called here at the end of each chunk.